### PR TITLE
Implement Video Recording in Envs Themselves

### PR DESCRIPTION
**Pull request by decodyng**
Per prior discussion, this PR makes it possible to specify a video save path when creating an env and use that to trigger videos being saved out from trajectories run on that env.

### CODE REVIEW

**Comment by brandonhoughton** [comment]
>

DEBUG:minerl.env.malmo.instance.3ea92e:[17:54:54] [Client thread/INFO]: [STDOUT]: Resized the window
--



DEBUG:minerl.env.malmo.instance.3ea92e:[17:54:54] [Client thread/INFO]: [STDOUT]: Unpausing
--



DEBUG:minerl.env.malmo.instance.3ea92e:[17:54:55] [Client thread/INFO]: [CHAT] §l400...
--



Traceback (most recent call last):
--



File "&lt;string&gt;", line 1, in &lt;module&gt;
--



File "/opt/conda/lib/python3.7/site-packages/gym/wrappers/time_limit.py", line 25, in reset
--



return self.env.reset(**kwargs)
--



File "/opt/conda/lib/python3.7/site-packages/minerl/env/_singleagent.py", line 23, in reset
--



multi_obs = super().reset()
--



File "/opt/conda/lib/python3.7/site-packages/minerl/env/_multiagent.py", line 502, in reset
--



return self._peek_obs()
--



File "/opt/conda/lib/python3.7/site-packages/minerl/env/_multiagent.py", line 685, in _peek_obs
--



self._reset_video_recorders(multi_obs[actor_name], actor_index)
--



File "/opt/conda/lib/python3.7/site-packages/minerl/env/_multiagent.py", line 514, in _reset_video_recorders
--



video_path = self.video_directory / f"{agent_ind}_{self._seed}_{round(time.time())}.mp4"
--



AttributeError: '_SingleAgentEnv' object has no attribute 'video_directory'
--



DEBUG:minerl.env.malmo.instance.3ea92e:[17:54:57] [Server thread/INFO]: [STDOUT]: Warning: managed 6/20 ticks this second.
--



INFO:process_watcher:About to reap process tree of 4791:launchClient.sh: i zombie, owner 4787, printing process tree status in termination order:
--



INFO:process_watcher:	-4791:launchClient.sh: i zombie, owner 4787
--



INFO:process_watcher:Trying to SIGTERM 4791:launchClient.sh: i zombie, owner 4787
--



INFO:process_watcher:Process psutil.Popen(pid=4791, status='terminated') terminated with exit code 1
--


<span class="term-fg31">🚨 Error: The command exited with status 1</span>

**Comment by brandonhoughton** [comment]
Tests are flaky btw because github has not auto-rebased this branch so it is running 2x instead of 1x test in parallel - I re-ran 2 times and it is working now!

File: minerl/env/_multiagent.py [hunk 1]
```diff
@@ -481,6 +513,14 @@ def reset(self) -> Any:
             # the episode in a cascading fashion
             self._seed = None
 
+    def _reset_video_recorders(self, obs, agent_ind):
+        # TODO figure out what other metadata we might want saved?
+        video_path = self.video_directory / f"{agent_ind}_{self._seed}_{round(time.time())}.mp4"
+        if self.video_writers[agent_ind].is_open():
+            self.video_writers[agent_ind].close()
+        self.video_writers[agent_ind].open(video_path)
```
Comment (brandonhoughton) [inline minerl/env/_multiagent.py:521]: It's hard to check via github but does this handle the case where I ask to record agents 3 and 4? Do we only call _reset_video_recorders for each self.record_agents?

**Review by brandonhoughton — APPROVED** [review]

File: minerl/env/_multiagent.py (hunk 1 above)
Comment (Miffyli) [reply minerl/env/_multiagent.py:521]: Yes, this is only called for agents that are in self.record_agents (indeces of agents to-be-recorded).

**Comment by Miffyli** [comment]
@decodyng Things look good here! I think we can keep this branch for BASALT evaluator package so you can close this PR

**Comment by shwang** [comment]
Eliminate the all option (because we aren't uploading this dataset for now)

**Comment by decodyng** [comment]
Closing this PR per reasoning given by @Miffyli

**Pull request closed without merging**